src/HDF5.jl

```diff
@@ -1181,16 +1181,25 @@ function Base.read(obj::DatasetOrAttribute, ::Type{T}, I...) where T
         end
     end
 
-    buf = Array{T}(undef, sz...)
-    memspace = dataspace(buf)
+    sz = Int.(sz)
+    if do_normalize(T)
+        buf = Matrix{UInt8}(undef, sizeof(T), prod(sz))
```

Comment:
We do `prod` instead of just keeping all the dimensional information here?

Reply:
I was motivated by trying to make sure the array comprehension didn't add too much overhead. I was hoping the `... ind in axes(buf, 2)` would optimize well across all versions of Julia, though I admit I haven't actually performance tested against `... ind in CartesianIndices(Base.tail(axes(buf)))`.

Reply:
Yeah, I'm guessing it's just too cumbersome to keep the dimension information. Is the thinking that the compilation time would be affected due to specialization on the dimension size in the `normalize_types` method? The Matrix keeps things simple in this stage, with the only negative that it requires a final reshape. 

I think it would be a good idea to add a comment above this line along the lines:
```
#  the data is stored in a buffer matrix where the first dimension at any stage are the bytes for a single data element
and the second-dimension of the matrix stores each element 
```

Reply:
> Is the thinking that the compilation time would be affected due to specialization on the dimension size in the normalize_types method? The Matrix keeps things simple in this stage, with the only negative that it requires a final reshape.

Yes, that was exactly my thinking. I have to admit I haven't actually extensively tested this assumption, though.

> I think it would be a good idea to add a comment above this line along the lines:

Sure, I'll get that added tomorrow.